pkg/virt-handler/vm.go

```diff
@@ -3005,150 +3000,243 @@ func (c *VirtualMachineController) vmUpdateHelperDefault(origVMI *v1.VirtualMach
 	if err != nil {
 		return err
 	}
+
 	var errorTolerantFeaturesError []error
 	disksInfo := map[string]*containerdisk.DiskInfo{}
-	if !vmi.IsRunning() && !vmi.IsFinal() {
-		// give containerDisks some time to become ready before throwing errors on retries
-		info := c.getLauncherClientInfo(vmi)
-		if ready, err := c.containerDiskMounter.ContainerDisksReady(vmi, info.NotInitializedSince); !ready {
-			if err != nil {
-				return err
-			}
-			c.queue.AddAfter(controller.VirtualMachineInstanceKey(vmi), time.Second*1)
-			return nil
-		}
+	readyToProceed, err := c.handleVMIState(vmi, cgroupManager, &disksInfo, &errorTolerantFeaturesError)
+	if err != nil {
+		return err
+	}
 
-		disksInfo, err = c.containerDiskMounter.MountAndVerify(vmi)
-		if err != nil {
-			return err
-		}
+	if !readyToProceed {
+		return nil
+	}
 
-		// Try to mount hotplug volume if there is any during startup.
-		if err := c.hotplugVolumeMounter.Mount(vmi, cgroupManager); err != nil {
-			return err
-		}
+	// Synchronize the VirtualMachineInstance state
+	err = c.syncVirtualMachine(client, vmi, preallocatedVolumes, disksInfo)
+	if err != nil {
+		return err
+	}
 
-		if err := c.setupNetwork(vmi, netsetup.FilterNetsForVMStartup(vmi)); err != nil {
-			return fmt.Errorf("failed to configure vmi network: %w", err)
-		}
+	// Post-sync housekeeping
+	err = c.handleHousekeeping(vmi, cgroupManager, domainExists)
+	if err != nil {
+		return err
+	}
 
-		isolationRes, err := c.podIsolationDetector.Detect(vmi)
-		if err != nil {
-			return fmt.Errorf(failedDetectIsolationFmt, err)
-		}
-		virtLauncherRootMount, err := isolationRes.MountRoot()
-		if err != nil {
-			return err
-		}
+	return errors.NewAggregate(errorTolerantFeaturesError)
+}
 
-		err = c.claimDeviceOwnership(virtLauncherRootMount, "kvm")
-		if err != nil {
-			return fmt.Errorf("failed to set up file ownership for /dev/kvm: %v", err)
-		}
-		if virtutil.IsAutoAttachVSOCK(vmi) {
-			if err := c.claimDeviceOwnership(virtLauncherRootMount, "vhost-vsock"); err != nil {
-				return fmt.Errorf("failed to set up file ownership for /dev/vhost-vsock: %v", err)
-			}
-		}
+func (c *VirtualMachineController) handleVMIState(vmi *v1.VirtualMachineInstance, cgroupManager cgroup.Manager, disksInfo *map[string]*containerdisk.DiskInfo, errorTolerantFeaturesError *[]error) (bool, error) {
+	if vmi.IsRunning() {
+		return true, c.handleRunningVMI(vmi, cgroupManager, errorTolerantFeaturesError)
+	} else if !vmi.IsFinal() {
+		return c.handleStartingVMI(vmi, cgroupManager, disksInfo)
+	}
+	return true, nil
+}
+
+func (c *VirtualMachineController) handleRunningVMI(vmi *v1.VirtualMachineInstance, cgroupManager cgroup.Manager, errorTolerantFeaturesError *[]error) error {
```

Comment:
The descriptions in the commit message about those 3 functions is really useful, maybe we can add them as comment? Like:
```suggestion
// handleRunningVMI contains the logic specifically for running VMs (hotplugging in running state, metrics, network updates)
func (c *VirtualMachineController) handleRunningVMI(vmi *v1.VirtualMachineInstance, cgroupManager cgroup.Manager, errorTolerantFeaturesError *[]error) error {
```

Reply:
Will add!